### PR TITLE
Finished and unit tested ParseArgs

### PR DESCRIPTION
**Pull request by alexander871015**
Split flag into flag.h and flag.cc so that it's easier for future use and prevent gflags from multiple definition.

### CODE REVIEW

File: hologram_dashboard/fetcher/flags.cc [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+*   limitations under the License.
+*/
+
+#include "flags.h"
+
+DEFINE_string(chipper_batch_job_cell, "", 
+    "The job running cell of Chipper Batch.");
+DEFINE_string(chipper_gdpr_batch_job_cell, "", 
+    "The job running cell of Chipper GDPR pipeline.");
+DEFINE_string(config_file_path, "",
```
Comment (sarahchen-google) [inline hologram_dashboard/fetcher/flags.cc:23]: The flag name is too generic, make it concrete.

File: hologram_dashboard/fetcher/hologram_data_source_availability_fetcher.cc [hunk 2]
```diff
@@ -34,4 +34,16 @@ void HologramDataSourceAvailabilityFetcher::
                                                 &hologram_configs_));
 }
 
+std::string HologramDataSourceAvailabilityFetcher::ParseFlags(int argc, char* argv[]) {
+    gflags::ParseCommandLineFlags(&argc, &argv, true);
```
Comment (sarahchen-google) [inline hologram_dashboard/fetcher/hologram_data_source_availability_fetcher.cc:38]: This should be done in main only.

File: hologram_dashboard/fetcher/hologram_data_source_availability_fetcher.h [hunk 3]
```diff
@@ -48,6 +41,12 @@ class HologramDataSourceAvailabilityFetcher: public HologramDataFetcher {
     // Poulates hologram_config_ and ends the program if the path provided leads
     // to wrong file or malformed file.
     void AcquireConfig(const std::string& config_file_path);
+
+    FRIEND_TEST(FetcherTest, MissingFlags);
+    FRIEND_TEST(FetcherTest, ValidFlags);
+    // Populates system_to_cell_map_ given the arguments and returns the path
+    // to the config file.
+    std::string ParseFlags(int argc, char* argv[]);
```
Comment (sarahchen-google) [inline hologram_dashboard/fetcher/hologram_data_source_availability_fetcher.h:49]: Move "int argc, char* argv[]" to main.

File: hologram_dashboard/fetcher/flags.h [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+*   See the License for the specific language governing permissions and
+*   limitations under the License.
+*/
+
+#pragma once
+#include <gflags/gflags.h>
+
+DECLARE_string(chipper_batch_job_cell);
+DECLARE_string(chipper_gdpr_batch_job_cell);
+DECLARE_string(config_file_path);
```
Comment (sarahchen-google) [inline hologram_dashboard/fetcher/flags.h:22]: I found ABSL_FLAG is public lib.

Can you switch to ABSL_FLAG? Its usage is very similar to this one.

https://abseil.io/docs/cpp/guides/flags

File: hologram_dashboard/fetcher/hologram_data_source_availability_fetcher_test.cc [hunk 5]
```diff
@@ -46,6 +46,49 @@ TEST(FetcherTest, ValidAcquireConfig) {
...
+    // Missing one flag.
+    int argc = 3;
+    // Allocate space for each of the arguments.
+    char arg1[35];
+    char arg2[35];
+    char arg3[35];
+    char* argv[3] = {arg1, arg2, arg3};
+    sprintf(arg1, "fetcher_main");
+    sprintf(arg2, "-chipper_batch_job_cell=cv");
+    sprintf(arg3, "-config_file_path=path");
```
Comment (sarahchen-google) [inline hologram_dashboard/fetcher/hologram_data_source_availability_fetcher_test.cc:60]: You can use 

absl::SetFlag(FLAGS_chipper_batch_job_cell)

File: hologram_dashboard/fetcher/flags.h (hunk 4 above)
Comment (alexander871015) [reply hologram_dashboard/fetcher/flags.h:22]: is there a reason why we would want to switch to absl_flag over gflags?

File: hologram_dashboard/fetcher/flags.cc (hunk 1 above)
Comment (alexander871015) [reply hologram_dashboard/fetcher/flags.cc:23]: what do you mean by too generic could you give me an example of a more concrete flag name?

File: hologram_dashboard/fetcher/flags.h (hunk 4 above)
Comment (sarahchen-google) [reply hologram_dashboard/fetcher/flags.h:22]: gflags are used ~2 years ago. Now all Google codes are switching to use ABSL_FLAG. We should keep in sync.

Originally I thought ABSL_FLAG is not a public lib. My bad.

File: hologram_dashboard/fetcher/hologram_data_source_availability_fetcher_test.cc [hunk 6]
```diff
@@ -46,6 +46,49 @@ TEST(FetcherTest, ValidAcquireConfig) {
...
+    EXPECT_EQ("path", hologram_fetcher.ParseFlags(argc, argv));
+    for (auto it = hologram_fetcher.system_to_cell_map_.begin();
+        it != hologram_fetcher.system_to_cell_map_.end(); it++) {
+        EXPECT_TRUE(it->first == "CHIPPER" || it->first == "CHIPPER_GDPR");
+        if (it->first == "CHIPPER") {
+            EXPECT_EQ(it->second, "cv");
+        }
+        else {
+            EXPECT_EQ(it->second, "ef");
+        }
```
Comment (emilyqu) [inline hologram_dashboard/fetcher/hologram_data_source_availability_fetcher_test.cc:88]: Can you do 
EXPECT_THAT(hologram_fetcher.system_to_cell_map_, UnorderedElementsAre(Pair("CHIPPER", "cv"), Pair(...)));?

File: hologram_dashboard/fetcher/hologram_data_source_availability_fetcher.cc [hunk 7]
```diff
@@ -34,4 +34,17 @@ void HologramDataSourceAvailabilityFetcher::
                                                 &hologram_configs_));
 }
 
+std::string HologramDataSourceAvailabilityFetcher::ParseFlags() {
+    // Ensure that all the flags are given.
+    assert(absl::GetFlag(FLAGS_chipper_batch_job_cell) != "");
```
Comment (sarahchen-google) [inline hologram_dashboard/fetcher/hologram_data_source_availability_fetcher.cc:39]: !absl::GetFlag(FLAGS_chipper_batch_job_cell).empty()

File: hologram_dashboard/fetcher/hologram_data_source_availability_fetcher.cc [hunk 8]
```diff
@@ -34,4 +34,17 @@ void HologramDataSourceAvailabilityFetcher::
...
+    assert(absl::GetFlag(FLAGS_chipper_gdpr_batch_job_cell) != "");
```
Comment (sarahchen-google) [inline hologram_dashboard/fetcher/hologram_data_source_availability_fetcher.cc:40]: Same above.

File: hologram_dashboard/fetcher/hologram_data_source_availability_fetcher.cc [hunk 9]
```diff
@@ -34,4 +34,17 @@ void HologramDataSourceAvailabilityFetcher::
...
+    assert(absl::GetFlag(FLAGS_hologram_source_config_file_path) != "");
```
Comment (sarahchen-google) [inline hologram_dashboard/fetcher/hologram_data_source_availability_fetcher.cc:41]: Same above.

File: hologram_dashboard/fetcher/hologram_data_source_availability_fetcher.cc [hunk 10]
```diff
@@ -34,4 +34,17 @@ void HologramDataSourceAvailabilityFetcher::
...
+
+    system_to_cell_map_["CHIPPER"] =
```
Comment (sarahchen-google) [inline hologram_dashboard/fetcher/hologram_data_source_availability_fetcher.cc:43]: Add proto enum to declear CHIPPER and CHIPPER_GDPR.

File: hologram_dashboard/fetcher/hologram_data_source_availability_fetcher.cc [hunk 11]
```diff
@@ -34,4 +34,17 @@ void HologramDataSourceAvailabilityFetcher::
...
+    // Ensure that all the flags are given.
+    assert(absl::GetFlag(FLAGS_chipper_batch_job_cell) != "");
+    assert(absl::GetFlag(FLAGS_chipper_gdpr_batch_job_cell) != "");
+    assert(absl::GetFlag(FLAGS_hologram_source_config_file_path) != "");
+
+    system_to_cell_map_["CHIPPER"] = 
+        absl::GetFlag(FLAGS_chipper_batch_job_cell);
+    system_to_cell_map_["CHIPPER_GDPR"] = 
+        absl::GetFlag(FLAGS_chipper_gdpr_batch_job_cell);
+    return absl::GetFlag(FLAGS_hologram_source_config_file_path);
```
Comment (sarahchen-google) [inline hologram_dashboard/fetcher/hologram_data_source_availability_fetcher.cc:47]: Why are you returning a flag value as output for this function?

File: hologram_dashboard/fetcher/hologram_data_source_availability_fetcher.cc (hunk 7 above)
Comment (alexander871015) [reply hologram_dashboard/fetcher/hologram_data_source_availability_fetcher.cc:39]: Can you show me where's the api for this function call? I can't seem to find one

File: hologram_dashboard/fetcher/hologram_data_source_availability_fetcher.cc (hunk 11 above)
Comment (alexander871015) [reply hologram_dashboard/fetcher/hologram_data_source_availability_fetcher.cc:47]: It's so that I can feed the file path into AcquireConfig

Comment (emilyqu) [reply hologram_dashboard/fetcher/hologram_data_source_availability_fetcher.cc:47]: It is better to parse the flag when you use it. no need to let a function return a flag value.

File: hologram_dashboard/fetcher/hologram_data_source_availability_fetcher_test.cc [hunk 12]
```diff
@@ -46,6 +46,36 @@ TEST(FetcherTest, ValidAcquireConfig) {
...
+TEST(FetcherTest, MissingFlags) {
+    HologramDataSourceAvailabilityFetcher hologram_fetcher;
+    // Missing one flag.
+    absl::SetFlag(&FLAGS_chipper_batch_job_cell, "ja");
+    absl::SetFlag(&FLAGS_hologram_source_config_file_path, "path");
+    ASSERT_DEATH(hologram_fetcher.ParseFlags(), "");
+    // Missing different flag.
+    absl::SetFlag(&FLAGS_chipper_gdpr_batch_job_cell, "cv");
+    absl::SetFlag(&FLAGS_hologram_source_config_file_path, "");
+    ASSERT_DEATH(hologram_fetcher.ParseFlags(), "");
```
Comment (emilyqu) [inline hologram_dashboard/fetcher/hologram_data_source_availability_fetcher_test.cc:58]: One test should only do one thing. Here you can just test if one flag is missing.

File: hologram_dashboard/fetcher/hologram_data_source_availability_fetcher_test.cc (hunk 6 above)
Comment (emilyqu) [reply hologram_dashboard/fetcher/hologram_data_source_availability_fetcher_test.cc:75]: is it resolved? in tests, we prefer not to use for loops.

File: hologram_dashboard/fetcher/hologram_data_source_availability_fetcher.cc (hunk 11 above)
Comment (alexander871015) [reply hologram_dashboard/fetcher/hologram_data_source_availability_fetcher.cc:47]: Should I change it to only asserting the flag is there, but not return the flag then?

Comment (sarahchen-google) [reply hologram_dashboard/fetcher/hologram_data_source_availability_fetcher.cc:47]: Can you grab some time with us for discussion, if you are confused in this part.

File: hologram_dashboard/fetcher/hologram_data_source_availability_fetcher.cc (hunk 7 above)
Comment (alexander871015) [reply hologram_dashboard/fetcher/hologram_data_source_availability_fetcher.cc:39]: Sorry I thought you were referring a special function that checks if a flag is empty, I have changed it accordingly

File: hologram_dashboard/fetcher/hologram_data_source_availability_fetcher.h [hunk 13]
```diff
@@ -27,32 +24,32 @@
...
+    FRIEND_TEST(FetcherTest, ConstructorMissingFlags);
+    FRIEND_TEST(FetcherTest, ConstructorValidFlags);
     FRIEND_TEST(FetcherTest, InvalidAcquireConfig);
     FRIEND_TEST(FetcherTest, ValidAcquireConfig);
     // Poulates hologram_config_ and ends the program if the path provided leads
     // to wrong file or malformed file.
     void AcquireConfig(const std::string& config_file_path);
     
-    std::unordered_map<std::string, std::string> system_to_cell_map_;
+    absl::flat_hash_map<System, std::string> system_to_cell_map_;
```
Comment (sarahchen-google) [inline hologram_dashboard/fetcher/hologram_data_source_availability_fetcher.h:49]: system_to_job_running_cell_

I don't put "map" as suffix, as it is already clear by variable type. But you can make the call.

File: hologram_dashboard/fetcher/proto/enums.proto [hunk 14]
```diff
@@ -65,4 +65,12 @@ enum DataSource {
     HOLOGRAM_USER_ATTRIBUTE = 5;
     HOLOGRAM_USER_EXPERIMENT_ID_AND_TEST_CODE = 6;
 }
+
+// Specify the system.
```
Comment (sarahchen-google) [inline hologram_dashboard/fetcher/proto/enums.proto:69]: // Existing systems that uses Holgram data.

File: hologram_dashboard/fetcher/hologram_data_source_availability_fetcher.h (hunk 13 above)
Comment (alexander871015) [reply hologram_dashboard/fetcher/hologram_data_source_availability_fetcher.h]: ya that makes sense i removed it

**Pull request merged**